Prevent Profile Creation for Invalid Cable Length

jianyuewu opened the pull request:
Added a check in BufferMgrDynamic::allocateProfile to skip profile creation when the cable length is 0m.

<!--
Please make sure you have read and understood the contribution guildlines:
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

1. Make sure your commit includes a signature generted with `git commit -s`
2. Make sure your commit title follows the correct format: [component]: description
3. Make sure your commit message contains enough details about the change and related tests
4. Make sure your pull request adds related reviewers, asignees, labels

Please also provide the following information in this pull request:
-->

**What I did**
Prevent profile creation for invalid cable length.

**Why I did it**
Ensures that profiles are not created for invalid cable  lengths, preventing potential configuration errors.

**How I verified it**
Local test: "config interface cable-length Ethernet0 5m" and "buffershow -l" to see if 0m profile is created.

**Details if related**
With out this change, profile will be created, log is:
2025 Feb 24 07:17:38.926732 r-leopard-70 NOTICE swss#buffermgrd: :- allocateProfile: Creating new profile 'pg_lossless_200000_0m_profile'                                                                     
2025 Feb 24 07:17:38.928570 r-leopard-70 NOTICE swss#buffermgrd: :- allocateProfile: BUFFER_PROFILE pg_lossless_200000_0m_profile has been created successfully                                               
2025 Feb 24 07:17:38.938143 r-leopard-70 NOTICE swss#orchagent: :- processBufferProfile: Created buffer profile pg_lossless_200000_0m_profile with type BUFFER_PROFILE_TABLE                                  
2025 Feb 24 07:17:38.939078 r-leopard-70 NOTICE swss#orchagent: :- processPriorityGroup: Set buffer PG Ethernet0:3-4 to BUFFER_PROFILE_TABLE:pg_lossless_200000_0m_profile

jianyuewu committed:
Prevent lossless profile creation for 0m cable

Changes:
- Skip PG creation when both cable length is 0m AND PG is lossless
- Allow lossy PGs to be created regardless of cable length

Test:
- Add unit test to verify lossy PG can be created with 0m cable length
- Verify lossy PG profile and attributes are set correctly

Signed-off-by: Jianyue Wu <jianyuew@nvidia.com>

baorliu committed:
[SWSS:portsorch] fix child_ports checking in addLagMember and removeLagMember for strip tag (#3343)

What I did
Added child_ports check in addLagMember and removeLagMember for strip tag

Why I did it
portorch sets LAG member's strip tag when adding subport:

    // Change hostif vlan tag for the parent port only when a first subport is created
    if (parentPort.m_child_ports.empty())
    {
        if (!setHostIntfsStripTag(parentPort, SAI_HOSTIF_VLAN_TAG_KEEP))
but if a new member is added later, in addLagMember function, it does not handle strip tag anymore. Cause the new added lag member has wrong tag mode.

abdosi committed:
Added Change to Skip Route Programming if NH is link/oper down (#3520)

*What I did:
Added Change to Skip Route Programming if NH is link/oper down. With Scale Route testing of 60K+ routes when we toggle all the interfaces[14+ interface back to back] as done here: https://github.com/sonic-net/sonic-mgmt/blob/master/tests/snappi_tests/multidut/bgp/test_bgp_outbound_uplink_multi_po_flap.py we see because of slowness of FRR Route APP_DB processing compare to Link Notification Handling where we have updated the Nexthop Group as part of Link Notification handling to point to default route via #3389 [if eligible] FRR slowness can reprogram the Route back to Nexthop which is link down.

This change is similar to #3394 which was done for Nexthop Group.

kperumalbfn committed:
Set Port UPDATE_DSCP attribute when TC_TO_DSCP map is attached (#3517)

* Set Port UPDATE_DSCP attribute when TC_TO_DSCP map is attached
What I did

Set Port SAI attribute SAI_PORT_ATTR_UPDATE_DSCP when TC_TO_DSCP map is attached to the port.
Why I did it

Some vendor SAI expects Sonic to set this attribute explicitly when TC_TO_DSCP map is attached to the port to modify DSCP value of the packet.

jianyuewu commented:
Local mgmt test cases all passed: https://allure.nvidia.com/allure-docker-service/projects/sonic-lionfish-r-lionfish-07-qos-test-buffer-py/reports/549/index.html

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1471,47 @@ namespace buffermgrdyn_test
+        InitBufferPg("Ethernet0|3-4");
+
+        auto expectedProfile = "pg_lossless_100000_0m_profile";
+        auto &portPgMap = m_dynamicBuffer->m_bufferProfileLookup[expectedProfile].port_pgs;
```
This will create an empty profile by using [] operator

jianyuewu replied:
Thank you. Yes, I'll use at() instead.

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -2932,6 +2932,12 @@ task_process_status BufferMgrDynamic::handleSingleBufferPgEntry(const string &ke
 
+    if (portInfo.cable_length == "0m")
+    {
+        SWSS_LOG_NOTICE("Skipping profile creation for cable length '0m'");
```
If I configure it to 0m and then add a PG, the PG configuration is lost.
When you change the cable length to other values, no profile will be created

jianyuewu replied:
OK understand, I'll move it to after allocateProfile() API create profile

stephenxs commented:
I believe we also need to add a test case to cover the scenario that it changes between 0m and other values

mukeshmv committed:
Add appliance entry validation (#3494)

* Add appliance entry validation (#3494)
- Do not allow more than 1 entry in DASH Appliance table.
- Do not allow DASH VNET creation before DASH Appliance entry creation.
- DASH ENI already has similar check for Appliance entry.

vivekrnv committed:
[smartswitch] Add support for ENI Based Forwarding (#3398)

* [smartswitch] Add support for ENI Based Forwarding
HLD: sonic-net/SONiC#1842
Requires sonic-net/sonic-swss-common#976
Add DashEniFwdOrch which installs ACL rules to Redirect the DASH packet to corresponding DPU

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1471,77 @@ namespace buffermgrdyn_test
+        InitCableLength("Ethernet0", "0m");
+        m_dynamicBuffer->doTask(m_selectableTable);
+        // Check that the profiles for 0m do not exist
+        ASSERT_TRUE(m_dynamicBuffer->m_bufferProfileLookup.find("pg_lossless_100000_0m_profile") == m_dynamicBuffer->m_bufferProfileLookup.end());
```
the 5m profile should also be removed, right?

jianyuewu commented:
> I believe we also need to add a test case to cover the scenario that it changes between 0m and other values

Thank you~ test case updated.
SONiC CI also passed with latest code update: https://nbuprod.blsm.nvidia.com/nbu-sws-sonic/job/sonic_ci/558/

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1471,77 @@ namespace buffermgrdyn_test
+        // 4. Update cable length to 5m, then Ethernet0:6 and Ethernet0:3-4 should be created with 5m profile
+        InitCableLength("Ethernet0", "5m");
+        // Simulate the task to apply changes
+        m_dynamicBuffer->doTask(m_selectableTable);
```
```suggestion
        m_dynamicBuffer->doTask();
```

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1471,77 @@ namespace buffermgrdyn_test
+
+        // 5. Update cable length to 0m, then Ethernet0:6 should be deleted, Ethernet0:3-4 should be deleted and 0m profile also not exist, and 5m profile should be deleted, PG also deleted, as profiles not exist
+        InitCableLength("Ethernet0", "0m");
+        m_dynamicBuffer->doTask(m_selectableTable);
```
```suggestion
        m_dynamicBuffer->doTask();
```

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1440,6 +1440,23 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
+        {
+            if (!pg.second.running_profile_name.empty())
+            {
+                m_bufferProfileLookup.erase(pg.second.running_profile_name);
```
what happens if another port still references the profile?

prgeor committed:
Initialize Port oper error map only once (#3538)

What I did
Initialize the port error status map only once

Why I did it
Any change in CONFIG_DB PORT table was resulting in updating the port error status map resulting in change in error count = 0

How I verified it
Verified by raising RF and LF error events on the port

prabhataravind committed:
[copp]: Use non-zero trap priority for default trap group (#3502)

* Use non-zero trap priority for default trap group
This priority is used internally in some vendor SAI implementations and causes
undesirable packet trapping behavior.

How I verified it

By running copp tests which include rate-limiting tests for TTL 1 packets on on Cisco/Mellanox/Arista platforms.
By manual tests with TTL 1 packets generated by scapy

stephenxs committed:
Optimize counter initialization by reducing the number of bulk counter poll calls and communication between swss/sairedis (#3504)

* Use flex counter manager for the following counter groups

- priority group watermark
- priority group drop
- queue watermark
- port counter group

Signed-off-by: Stephen Sun <stephens@nvidia.com>

* Fix compiling error

Signed-off-by: Stephen Sun <stephens@nvidia.com>

* Support bulk create also for WRED/ECN counter groups

Signed-off-by: Stephen Sun <stephens@nvidia.com>

* Fix review comments

Signed-off-by: Stephen Sun <stephens@nvidia.com>

---------

Signed-off-by: Stephen Sun <stephens@nvidia.com>

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -944,6 +944,13 @@ void BufferMgrDynamic::updateBufferObjectListToDb(const string &key, const strin
 // We have to check the headroom ahead of applying them
 task_process_status BufferMgrDynamic::allocateProfile(const string &speed, const string &cable_len, const string &mtu, const string &threshold, const string &gearbox_model, long lane_count, string &profile_name)
 {
+    bool is_lossless = (profile_name.find("pg_lossless_") != string::npos);
```
Why is this required?

stephenxs commented:
> Local mgmt test cases all passed: https://allure.nvidia.com/allure-docker-service/projects/sonic-lionfish-r-lionfish-07-qos-test-buffer-py/reports/549/index.html

Do not unheal internal links, even in private PR. Please delete it.

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1462,6 +1469,19 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
                 continue;
             }
 
+            if (cable_length == "0m" && it->second.lossless)
```
let's use the reference to it->second defined in L1459.
```suggestion
            if (cable_length == "0m" && portPg.lossless)
```

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1462,6 +1469,19 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
+            if (cable_length == "0m" && it->second.lossless)
+            {
+                updateBufferObjectToDb(key, oldProfile, false); // 0m, remove PG.
+                SWSS_LOG_NOTICE("All profiles for port %s have been removed due to cable length being set to '0m'", port.c_str());
```
`All profiles`? we do not remove lossy profiles, right

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1462,6 +1469,19 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
 
+            if (cable_length == "0m" && it->second.lossless)
+            {
+                updateBufferObjectToDb(key, oldProfile, false); // 0m, remove PG.
```
what happens if there was no lossless profile? will there be an error message?

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1476,86 @@ namespace buffermgrdyn_test
         ASSERT_EQ(m_dynamicBuffer->m_portInfoLookup["Ethernet12"].state, PORT_READY);
     }
+
+    TEST_F(BufferMgrDynTest, SkipProfileCreationForZeroCableLength)
```
could you add some comments explaining what this test does?

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1476,86 @@ namespace buffermgrdyn_test
+
+        // 2. Check if the reference for 5m profile of this port had been deleted in buffer PG table
+        // Since the cable length is set to 0m, previous profile for 5m should not exist.
+        ASSERT_TRUE(m_dynamicBuffer->m_bufferProfileLookup.find("pg_lossless_100000_5m_profile") == m_dynamicBuffer->m_bufferProfileLookup.end());
```
where was the cable length set to 5m?

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1476,86 @@ namespace buffermgrdyn_test
+        // 6. Check if port MTU update, PG and profile still there
+        InitCableLength("Ethernet0", "5m");
+        string mtu = "4096";
+        m_dynamicBuffer->m_portInfoLookup["Ethernet0"].mtu = mtu;
```
why not setting it via config_db?

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -944,6 +944,13 @@ void BufferMgrDynamic::updateBufferObjectListToDb(const string &key, const strin
 // We have to check the headroom ahead of applying them
 task_process_status BufferMgrDynamic::allocateProfile(const string &speed, const string &cable_len, const string &mtu, const string &threshold, const string &gearbox_model, long lane_count, string &profile_name)
 {
+    bool is_lossless = (profile_name.find("lossless") != string::npos);
```
we should not rely on name. don't we have other field indicating lossless/lossy?

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1462,6 +1468,25 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
+                updateBufferObjectToDb(key, oldProfile, false);
+                SWSS_LOG_NOTICE("All profiles for port %s have been removed due to cable length being set to '0m'", port.c_str());
+
+                if (!oldProfile.empty())
```
we already tested `oldProfile.empty()` in L1474 and continue if it was.

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1462,6 +1468,25 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
+            {
+                if (oldProfile.empty())
+                {
+                    SWSS_LOG_WARN("No lossless profile found for port %s when cable length is set to '0m'.", port.c_str());
```
why warning?

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -944,6 +944,12 @@ void BufferMgrDynamic::updateBufferObjectListToDb(const string &key, const strin
 // We have to check the headroom ahead of applying them
 task_process_status BufferMgrDynamic::allocateProfile(const string &speed, const string &cable_len, const string &mtu, const string &threshold, const string &gearbox_model, long lane_count, string &profile_name)
 {
+    if (cable_len == "0m")
```
do we still need to test `cable_len` here? I suppose it's always false.

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1476,140 @@ namespace buffermgrdyn_test
+        HandleTable(cableLengthTable);
+        // Expect profile not created
+        auto zeroMProfile = "pg_lossless_100000_0m_profile";
+        ASSERT_EQ(m_dynamicBuffer->m_bufferProfileLookup.find(zeroMProfile), m_dynamicBuffer->m_bufferProfileLookup.end());
```
check no lossless PG is created for Ethernet0:3-4?

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1476,141 @@ namespace buffermgrdyn_test
+        // Expect profile not created
+        auto zeroMProfile = "pg_lossless_100000_0m_profile";
+        ASSERT_EQ(m_dynamicBuffer->m_bufferProfileLookup.find(zeroMProfile), m_dynamicBuffer->m_bufferProfileLookup.end());
+        ASSERT_TRUE(m_dynamicBuffer->m_portPgLookup.find("Ethernet0:3-4") == m_dynamicBuffer->m_portPgLookup.end());
```
can you check what the key of m_portPgLookup is?

jianyuewu committed:
Keep PGs, remove losslesss profiles when 0m cable

When 0m cable, lossy PG and profile can still be created,
while lossless profile will be deleted.

Signed-off-by: Jianyue Wu <jianyuew@nvidia.com>

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1462,6 +1462,31 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
+                    SWSS_LOG_NOTICE("Attempted to remove non-existent profile %s for port %s.", oldProfile.c_str(), port.c_str());
+                }
+                portPg.running_profile_name.clear();
+                SWSS_LOG_NOTICE("All lossless profiles and PGs for port %s will be removed due to cable length being set to '0m'",
```
This log is printed for many times.

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1462,6 +1462,31 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
+                }
+                else
+                {
+                    SWSS_LOG_NOTICE("Attempted to remove non-existent profile %s for port %s.", oldProfile.c_str(), port.c_str());
```
What does it mean?

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1476,137 @@ namespace buffermgrdyn_test
+                            });
+        HandleTable(cableLengthTable);
+        string mtu = "4096";
+        m_dynamicBuffer->m_portInfoLookup["Ethernet0"].mtu = mtu;
```
Do not directly change internal data. Push it to db and then call doTask

stephenxs commented on tests/mock_tests/buffermgrdyn_ut.cpp:
```diff
@@ -1471,4 +1476,137 @@ namespace buffermgrdyn_test
+        ASSERT_TRUE(m_dynamicBuffer->m_portPgLookup["Ethernet0"].find("Ethernet0:6") != m_dynamicBuffer->m_portPgLookup["Ethernet0"].end());
+
+        // Lossy profiles still there
+        ASSERT_EQ(m_dynamicBuffer->m_portPgLookup["Ethernet0"]["Ethernet0:0"].running_profile_name, "ingress_lossy_profile");
```
We should check lossy profile after each step.

stephenxs commented:
Do not suggest force pushing frequently. I can hardly figure out what has been changed between commits.
Thanks

jianyuewu committed:
Remove unnecessary prints and refactor UT

Signed-off-by: Jianyue Wu <jianyuew@nvidia.com>

jianyuewu commented:
> Do not suggest force pushing frequently. I can hardly figure out what has been changed between commits. Thanks

OK, understand, I'll use incremental commit instead😊

stephenxs commented on cfgmgr/buffermgrdyn.cpp:
```diff
@@ -1462,6 +1462,26 @@ task_process_status BufferMgrDynamic::refreshPgsForPort(const string &port, cons
+            {
+                if (oldProfile.empty())
+                {
+                    SWSS_LOG_NOTICE("No lossless profile found for port %s when cable length is set to '0m'.", port.c_str());
```
It isn't worth a notice message
```suggestion
                    SWSS_LOG_INFO("No lossless profile found for port %s when cable length is set to '0m'.", port.c_str());
```

jianyuewu committed:
Update log level to info

Co-authored-by: Stephen Sun <5379172+stephenxs@users.noreply.github.com>

jianyuewu committed:
Merge branch 'master' into skip_0m_profile

Pterosaur committed:
Update test_macsec.py (#3549)

What I did
Provide an explicit value for send_sci in the macsec vstest

Why I did it

The kernel 5.15 requires the send_sci to be true if the sci value was provided explicitly.

puffc committed:
[MCLAG] Fix a race condition when moving MAC addresses to MCLAG peer on one-arm MCLAG interface down. (#3524)

* [MCLAG] Fix a racce condition when moving MAC addresses to MCLAG peer on one-arm MCLAG interface down.
What I did
Avoid flushing MAC addresses for the MCLAG interface when the MCLAG interface goes down. This is because ICCPD will take care of those MAC addresses by moving them to the MCLAG peer interface.

Why I did it
Fix issue #2913 and also replace PR #2539.
When the MCLAG interface goes down, the iccpd will move the MAC addresses associated with this particular MCLAG interface to the peer link, according the MCLAG HLD:

theasianpianist committed:
Update gitignore for fabricmgrd, stpmgrd, and the p4orch_tests binaries. (#3552)

What I did
Add entries to the gitignore file for fabricmgrd, stpmgrd, and the p4orch_tests binaries.

Why I did it
When these binaries were added they were not included in the gitignore file. Adding them ensures that they cannot be accidentally committed.

dypet committed:
Use software_bfd instead of switch_type. (#3525)

* Use software_bfd instead of switch_type.
What I did
*Removed software BFD dependence on switch_type 'dpu'. Added checks 
 SAI_SWITCH_ATTR_SUPPORTED_IPV4_BFD_SESSION_OFFLOAD_TYPE and 
 SAI_SWITCH_ATTR_SUPPORTED_IPV6_BFD_SESSION_OFFLOAD_TYPE instead. 
*Will use software BFD if either return SAI_BFD_SESSION_OFFLOAD_TYPE_NONE

Why I did it
Request in #3406 to allow software BFD to be enabled independent of switch_type 'dpu'.

jianyuewu committed:
Merge branch 'master' into skip_0m_profile

PJHsieh committed:
[orchagent] Do not restore port admin if port admin is configured (#3447)

* [orchagent] Do not restore port admin if port admin is configured

Issue:
The pCfg.admin_status might be overridden by prior configurations.

Solution:
Only restore port admin if port admin is not configured.

jianyuewu committed:
Merge branch 'master' into skip_0m_profile
